src/render/expression.ts

```diff
@@ -39,12 +50,22 @@ export class Expression {
   }
 }
 
-export function evalToken (token: Token | undefined, ctx: Context): any {
+export function evalToken (token: Token | undefined, ctx: Context, lenient: boolean = false): any {
   assert(ctx, () => 'unable to evaluate: context not defined')
   if (TypeGuards.isPropertyAccessToken(token)) {
     const variable = token.getVariableAsText()
     const props: string[] = token.props.map(prop => evalToken(prop, ctx))
-    return ctx.get([variable, ...props])
+    try {
+      return ctx.get([variable, ...props])
+    } catch (e) {
+      // we catch the error thrown by Context.getFromScope() for undefined vars.
+      // Alt, we could make this more robust by setting a flag or using a separate error class.
+      if (lenient && e instanceof TypeError && e.message.startsWith("undefined variable:")) {
```

Comment:
So... This one was a bit tricky because `RenderError` takes a Template and `LiquidError` takes a Token, both of which we don't have available when the error is thrown in `Context.getFromScope()`. So I added two new error classes:
* `InternalUndefinedVariableError` is used to signal from `Context.getFromScope()` to `evalToken()`
* `evalToken()` would then (if any) enrich this error with its token information & re-raise an `UndefinedVariableError` subclass `LiquidError`.

Happy to learn if there's an easier method to get this right, but I didn't see any.